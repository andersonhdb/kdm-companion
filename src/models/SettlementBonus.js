import { types, getSnapshot } from 'mobx-state-tree'
import R from 'ramda'

export const SettlementBonus = types
  .model('SettlementBonus', {
    survival: types.maybe(types.number),
    accuracy: types.maybe(types.number),
    strength: types.maybe(types.number),
    evasion: types.maybe(types.number),
    insanity: types.maybe(types.number),
    courage: types.maybe(types.number),
    understanding: types.maybe(types.number),
    'hunt xp': types.maybe(types.number),
    description: types.maybe(types.array(types.string)),
  })
  .actions(self => ({
    add(bonus) {
      self.survival += bonus.survival || 0
      self.accuracy += bonus.accuracy || 0
      self.strength += bonus.strength || 0
      self.evasion += bonus.evasion || 0
      self.insanity += bonus.insanity || 0
      self.courage += bonus.courage || 0
      self.understanding += bonus.understanding || 0
      self['hunt xp'] += bonus['hunt xp'] || 0

      if (bonus.description) {
        if (!self.description) {
          self.description = []
        }
        self.description = self.description.concat(bonus.description)
      }
    },
    remove(bonus) {
      self.survival -= bonus.survival || 0
      self.accuracy -= bonus.accuracy || 0
      self.strength -= bonus.strength || 0
      self.evasion -= bonus.evasion || 0
      self.insanity -= bonus.insanity || 0
      self.courage -= bonus.courage || 0
      self.understanding -= bonus.understanding || 0
      self['hunt xp'] -= bonus['hunt xp'] || 0

      self.description = R.filter(text => {
        return !bonus.description.includes(text)
      }, self.description)
    },
  }))

export const init = {
  survival: 0,
  accuracy: 0,
  strength: 0,
  evasion: 0,
  insanity: 0,
  courage: 0,
  understanding: 0,
  'hunt xp': 0,
  description: [],
}
