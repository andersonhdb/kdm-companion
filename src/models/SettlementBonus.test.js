import { getSnapshot } from 'mobx-state-tree'
import { SettlementBonus } from './SettlementBonus'

it('initializes', () => {
  let sb = SettlementBonus.create()

  expect(getSnapshot(sb)).toMatchSnapshot()
})

it('.add empty bonus', () => {
  let sb = SettlementBonus.create()
  sb.add({})

  expect(getSnapshot(sb)).toMatchSnapshot()
})

it('.add bonus', () => {
  let sb = SettlementBonus.create()
  sb.add({ strength: 1, description: ['test1', 'test2'] })

  expect(getSnapshot(sb)).toMatchSnapshot()
})

it('.remove empty bonus', () => {
  let sb = SettlementBonus.create()
  sb.remove({})

  expect(getSnapshot(sb)).toMatchSnapshot()
})

it('.remove empty bonus description', () => {
  let bonus = { strength: 1, description: [] }
  let sb = SettlementBonus.create()
  sb.add(bonus)
  sb.remove(bonus)

  expect(getSnapshot(sb)).toMatchSnapshot()
})

it('.remove bonus', () => {
  let sb = SettlementBonus.create()
  sb.add({ strength: 1, description: ['test1', 'test2'] })
  sb.remove({ strength: 1, description: ['test1', 'test2'] })

  expect(getSnapshot(sb)).toMatchSnapshot()
})

it('.remove one bonus', () => {
  let sb = SettlementBonus.create()
  sb.add({ strength: 1, description: ['test1', 'test2'] })
  sb.add({ strength: 1, evasion: 1, description: ['test3'] })
  sb.remove({ strength: 1, description: ['test1', 'test2'] })

  expect(getSnapshot(sb)).toMatchSnapshot()
})
