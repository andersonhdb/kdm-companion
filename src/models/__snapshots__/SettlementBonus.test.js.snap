// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.add bonus 1`] = `
Object {
  "accuracy": 0,
  "courage": 0,
  "description": Array [
    "test1",
    "test2",
  ],
  "evasion": 0,
  "hunt xp": 0,
  "insanity": 0,
  "strength": 1,
  "survival": 0,
  "understanding": 0,
}
`;

exports[`.add empty bonus 1`] = `
Object {
  "accuracy": 0,
  "courage": 0,
  "description": Array [],
  "evasion": 0,
  "hunt xp": 0,
  "insanity": 0,
  "strength": 0,
  "survival": 0,
  "understanding": 0,
}
`;

exports[`.remove bonus 1`] = `
Object {
  "accuracy": 0,
  "courage": 0,
  "description": Array [],
  "evasion": 0,
  "hunt xp": 0,
  "insanity": 0,
  "strength": 0,
  "survival": 0,
  "understanding": 0,
}
`;

exports[`.remove empty bonus 1`] = `
Object {
  "accuracy": 0,
  "courage": 0,
  "description": Array [],
  "evasion": 0,
  "hunt xp": 0,
  "insanity": 0,
  "strength": 0,
  "survival": 0,
  "understanding": 0,
}
`;

exports[`.remove empty bonus description 1`] = `
Object {
  "accuracy": 0,
  "courage": 0,
  "description": Array [],
  "evasion": 0,
  "hunt xp": 0,
  "insanity": 0,
  "strength": 0,
  "survival": 0,
  "understanding": 0,
}
`;

exports[`.remove one bonus 1`] = `
Object {
  "accuracy": 0,
  "courage": 0,
  "description": Array [
    "test3",
  ],
  "evasion": 1,
  "hunt xp": 0,
  "insanity": 0,
  "strength": 1,
  "survival": 0,
  "understanding": 0,
}
`;

exports[`initializes 1`] = `
Object {
  "accuracy": 0,
  "courage": 0,
  "description": Array [],
  "evasion": 0,
  "hunt xp": 0,
  "insanity": 0,
  "strength": 0,
  "survival": 0,
  "understanding": 0,
}
`;
